File: app/src/main/java/com/bookyrself/bookyrself/views/EventDetailActivity.java
package com.bookyrself.bookyrself.views;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.util.Pair;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.CardView;
import android.support.v7.widget.Toolbar;
import android.text.Html;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.bookyrself.bookyrself.R;
import com.bookyrself.bookyrself.data.ResponseModels.EventDetail.EventDetail;
import com.bookyrself.bookyrself.data.ResponseModels.EventDetail.Host;
import com.bookyrself.bookyrself.data.ResponseModels.EventDetail.MiniUser;
import com.bookyrself.bookyrself.presenters.EventDetailPresenter;
import com.bookyrself.bookyrself.utils.CircleTransform;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.squareup.picasso.Picasso;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by benmedcalf on 11/22/17.
 */

public class EventDetailActivity extends AppCompatActivity implements EventDetailPresenter.EventDetailPresenterListener {

    @BindView(R.id.event_detail_date)
    TextView DateView;
    @BindView(R.id.item_event_detail_username)
    TextView HostUsernameTextview;
    @BindView(R.id.item_event_detail_citystate)
    TextView EventCityState;
    @BindView(R.id.item_event_detail_url)
    TextView HostUrlTextView;
    @BindView(R.id.event_detail_host_item)
    CardView HostcardView;
    @BindView(R.id.event_detail_toolbar)
    Toolbar Toolbar;
    @BindView(R.id.item_event_detail_userthumb)
    ImageView hostImageView;
    @BindView(R.id.event_detail_users_list)
    ListView usersListView;
    @BindView(R.id.event_detail_empty_state)
    View emptyState;
    @BindView(R.id.event_detail_linearlayout)
    View eventDetailContent;
    @BindView(R.id.empty_state_text_header)
    TextView emptyStateTextHeader;
    @BindView(R.id.empty_state_text_subheader)
    TextView emptyStateTextSubHeader;
    @BindView(R.id.empty_state_image)
    ImageView emptyStateImage;
    @BindView(R.id.empty_state_button)
    Button emptyStateButton;
    @BindView(R.id.event_detail_progressBar)
    ProgressBar progressBar;

    private EventDetailPresenter presenter;
    private StorageReference storageReference;
    private List<Pair<String, MiniUser>> invitedUsers;
    private UsersListAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_event_detail);
        ButterKnife.bind(this);

        // Get the storage reference for the thumbnails
        storageReference = FirebaseStorage.getInstance().getReference();

        // Set up the adapter
        invitedUsers = new ArrayList<>();
        adapter = new UsersListAdapter(this, invitedUsers, storageReference);
        usersListView.setAdapter(adapter);

        // Set up initial loading state
        eventDetailContent.setVisibility(View.GONE);
        emptyState.setVisibility(View.GONE);
        String eventId = getIntent().getStringExtra("eventId");
        setSupportActionBar(Toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        // Set up the presenter
        presenter = new EventDetailPresenter(this, eventId);
        presenter.subscribe();
    }

    @Override
    public void showEventData(final EventDetail eventDetail) {

        showProgressbar(false);
        Toolbar.setTitle(eventDetail.getEventname());


        // Set up the host card
        Host host = eventDetail.getHost();
        String hostUsername = host.getUsername();
        // Show the City State of the event, not the host
        String hostCityState = eventDetail.getCitystate();

        HostcardView.setOnClickListener(v -> {
            Intent intent = new Intent(HostcardView.getContext(), UserDetailActivity.class);
            intent.putExtra("userId", eventDetail.getHost().getUserId());
            startActivity(intent);
        });

        HostUsernameTextview.setText(hostUsername);
        final StorageReference profileImageReference = storageReference.child("images/" + host.getUserId());
        profileImageReference.getDownloadUrl().addOnSuccessListener(uri -> Picasso.with(getApplicationContext())
                .load(uri)
                .resize(148, 148)
                .centerCrop()
                .transform(new CircleTransform())
                .into(hostImageView)).addOnFailureListener(exception -> {
            // Handle any errors
            hostImageView.setImageDrawable(getApplicationContext().getDrawable(R.drawable.ic_profile_black_24dp));
        });


        // Set the date
        DateFormat inputformat = new SimpleDateFormat("yyyy-MM-dd", Locale.US);
        try {
            Date date = inputformat.parse(eventDetail.getDate());
            DateFormat outputFormat = new SimpleDateFormat("EEEE, MMMM dd, yyyy", Locale.US);
            String formattedDate = outputFormat.format(date);
            DateView.setText(formattedDate);
        } catch (ParseException e) {
            e.printStackTrace();
        }


        // Set the city state
        EventCityState.setText(hostCityState);

        // Make it visible
        eventDetailContent.setVisibility(View.VISIBLE);
    }

    @Override
    public void showInvitedUser(Pair<String, MiniUser> user) {
        invitedUsers.add(user);
        adapter.notifyDataSetChanged();
    }

    public void showProgressbar(Boolean show) {
        if (show) {
            progressBar.setVisibility(View.VISIBLE);
        } else {
            progressBar.setVisibility(View.GONE);
        }
    }

    @Override
    public void presentError(String message) {
        showProgressbar(false);
        Toolbar.setTitle("Event Detail Error");
        eventDetailContent.setVisibility(View.GONE);
        emptyStateButton.setVisibility(View.GONE);
        emptyStateImage.setImageDrawable(getDrawable(R.drawable.ic_error_empty_state));
        emptyStateTextHeader.setText(getString(R.string.error_header));
        emptyStateTextSubHeader.setText(message);
        emptyStateTextHeader.setVisibility(View.VISIBLE);
        emptyStateTextSubHeader.setVisibility(View.VISIBLE);
        emptyState.setVisibility(View.VISIBLE);
    }


    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    /**
     * Adapter
     */
    private class UsersListAdapter extends BaseAdapter {

        private final StorageReference mStorageReference;
        private Context mContext;
        private LayoutInflater mInflater;


        private UsersListAdapter(Context context, List<Pair<String, MiniUser>> miniUsers, StorageReference storageReference) {
            invitedUsers = miniUsers;
            mInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            mContext = context;
            mStorageReference = storageReference;
        }

        @Override
        public int getCount() {
            return invitedUsers.size();
        }

        @Override
        public Object getItem(int position) {
            return invitedUsers.get(position).second;
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            View rowView = mInflater.inflate(R.layout.item_event_detail_user, parent, false);
            final ImageView userThumb = rowView.findViewById(R.id.item_event_detail_userthumb);
            TextView userName = rowView.findViewById(R.id.item_event_detail_username);
            TextView cityState = rowView.findViewById(R.id.item_event_detail_citystate);
            TextView userUrl = rowView.findViewById(R.id.item_event_detail_url);
            TextView attendingStatusTextView = rowView.findViewById(R.id.item_event_detail_attending_textview);

            final MiniUser miniUser = (MiniUser) getItem(position);

            userName.setText(miniUser.getUsername());
            cityState.setText(miniUser.getCitystate());
            attendingStatusTextView.setText(miniUser.getAttendingStatus());
            userUrl.setClickable(true);
            userUrl.setMovementMethod(LinkMovementMethod.getInstance());
            String linkedText =
                    String.format("<a href=\"%s\">%s</a> ", ("http://" + miniUser.getUrl()), miniUser.getUrl());
            userUrl.setText(Html.fromHtml(linkedText));

            final StorageReference profileImageReference = mStorageReference.child("images/" + miniUser.getUserId());
            profileImageReference.getDownloadUrl().addOnSuccessListener(uri ->
                    Picasso.with(mContext)
                            .load(uri)
                            .resize(148, 148)
                            .centerCrop()
                            .transform(new CircleTransform())
                            .into(userThumb)).addOnFailureListener(exception ->
                    // Handle any errors
                    userThumb.setImageDrawable(mContext.getDrawable(R.drawable.ic_profile_black_24dp)));

            rowView.setOnClickListener(v -> {
                Intent intent = new Intent(mContext, UserDetailActivity.class);
                intent.putExtra("userId", miniUser.getUserId());
                mContext.startActivity(intent);
            });

            return rowView;
        }
    }
}

File: app/src/main/java/com/bookyrself/bookyrself/views/EventDetailActivity.kt
package com.bookyrself.bookyrself.views

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.util.Linkify
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import com.bookyrself.bookyrself.R
import com.bookyrself.bookyrself.data.ServerModels.EventDetail.EventDetail
import com.bookyrself.bookyrself.data.ServerModels.EventDetail.MiniUser
import com.bookyrself.bookyrself.presenters.EventDetailPresenter
import com.bookyrself.bookyrself.utils.CircleTransform
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.activity_event_detail.*
import kotlinx.android.synthetic.main.empty_state_template.*
import kotlinx.android.synthetic.main.item_event_detail_user.*
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by benmedcalf on 11/22/17.
 */

class EventDetailActivity : AppCompatActivity(), EventDetailPresenter.EventDetailPresenterListener {

    private var presenter: EventDetailPresenter? = null
    private var storageReference: StorageReference? = null
    private var invitedUsers: MutableList<Map.Entry<String, MiniUser>>? = null
    private var adapter: UsersListAdapter? = null
    private var eventId: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_event_detail)

        // Get the storage reference for the thumbnails
        storageReference = FirebaseStorage.getInstance().reference

        // Set up the adapter
        invitedUsers = ArrayList()
        adapter = UsersListAdapter(this, invitedUsers!!, storageReference!!)
        event_detail_users_list.adapter = adapter

        // Set up initial loading state
        event_detail_linearlayout!!.visibility = View.GONE
        event_detail_empty_state.visibility = View.GONE
        eventId = intent.getStringExtra("eventId")
        setSupportActionBar(event_detail_toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        showProgressbar(true)

        // Set up the presenter
        presenter = EventDetailPresenter(this, eventId!!)
        presenter!!.subscribe()
    }

    override fun showEventData(eventDetail: EventDetail) {

        showProgressbar(false)
        event_detail_collapsing_toolbar!!.title = eventDetail.eventname
        event_detail_collapsing_toolbar!!.setExpandedTitleColor(resources.getColor(R.color.cardview_light_background))
        event_detail_collapsing_toolbar!!.setCollapsedTitleTextColor(resources.getColor(R.color.cardview_light_background))

        // Set up the host card
        val host = eventDetail.host
        val hostUsername = host.username

        // Show the City State of the event, not the host
        val hostCityState = eventDetail.citystate

        event_detail_host_item!!.setOnClickListener {
            val intent = Intent(event_detail_host_item!!.context, UserDetailActivity::class.java)
            intent.putExtra("userId", eventDetail.host.userId)
            startActivity(intent)
        }

        item_event_detail_username!!.text = hostUsername
        val profileImageReference = storageReference!!.child("images/users/" + host.userId)
        profileImageReference.downloadUrl.addOnSuccessListener { uri ->
            Picasso.with(applicationContext)
                    .load(uri)
                    .resize(200, 200)
                    .centerCrop()
                    .transform(CircleTransform())
                    .into(item_event_detail_userthumb)
        }.addOnFailureListener {
            // Handle any errors
            item_event_detail_userthumb!!.setImageDrawable(applicationContext.getDrawable(R.drawable.ic_profile_black_24dp))
        }


        // Set the date
        val inputFormat = SimpleDateFormat("yyyy-MM-dd", Locale.US)
        try {
            val date = inputFormat.parse(eventDetail.date)
            val outputFormat = SimpleDateFormat("EEEE, MMMM dd, yyyy", Locale.US)
            val formattedDate = outputFormat.format(date)
            event_detail_date!!.text = formattedDate
        } catch (e: ParseException) {
            e.printStackTrace()
        }


        // Set the city state
        item_event_detail_citystate!!.text = hostCityState

        // Set the image
        val eventImageReference = storageReference!!.child("images/events/" + eventId!!)
        eventImageReference.downloadUrl.addOnSuccessListener { uri ->
            Picasso.with(applicationContext)
                    .load(uri)
                    .resize(500, 500)
                    .centerCrop()
                    .into(event_image_detail)
        }.addOnFailureListener { Log.e(this.javaClass.name, "Unable to load Event Image") }

        // Make it visible
        event_detail_linearlayout!!.visibility = View.VISIBLE
    }

    override fun showInvitedUser(user: AbstractMap.SimpleEntry<String, MiniUser>) {
        invitedUsers!!.add(user)
        adapter!!.notifyDataSetChanged()
    }

    private fun showProgressbar(show: Boolean) {
        if (show) {
            event_detail_progressBar!!.visibility = View.VISIBLE
        } else {
            event_detail_progressBar!!.visibility = View.GONE
        }
    }

    override fun presentError(message: String) {
        showProgressbar(false)
        event_detail_toolbar!!.title = "Event Detail Error"
        event_detail_linearlayout!!.visibility = View.GONE
        empty_state_button!!.visibility = View.GONE
        empty_state_image!!.setImageDrawable(getDrawable(R.drawable.ic_error_empty_state))
        empty_state_text_header!!.text = getString(R.string.error_header)
        empty_state_text_subheader!!.text = message
        empty_state_text_header!!.visibility = View.VISIBLE
        empty_state_text_subheader!!.visibility = View.VISIBLE
        event_detail_empty_state!!.visibility = View.VISIBLE
    }


    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    /**
     * Adapter
     */
    private inner class UsersListAdapter constructor(private val mContext: Context, miniUsers: MutableList<Map.Entry<String, MiniUser>>, private val mStorageReference: StorageReference) : BaseAdapter() {

        private val mInflater: LayoutInflater


        init {
            invitedUsers = miniUsers
            mInflater = mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        }

        override fun getCount(): Int {
            return invitedUsers!!.size
        }

        override fun getItem(position: Int): Any {
            return invitedUsers!![position].value
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val rowView = mInflater.inflate(R.layout.item_event_detail_user, parent, false)
            val userThumb = rowView.findViewById<ImageView>(R.id.item_event_detail_userthumb)
            val userName = rowView.findViewById<TextView>(R.id.item_event_detail_username)
            val cityState = rowView.findViewById<TextView>(R.id.item_event_detail_citystate)
            val userUrl = rowView.findViewById<TextView>(R.id.item_event_detail_url)
            val attendingStatusTextView = rowView.findViewById<TextView>(R.id.item_event_detail_attending_textview)

            val miniUser = getItem(position) as MiniUser

            userName.text = miniUser.username
            cityState.text = miniUser.citystate
            attendingStatusTextView.text = miniUser.attendingStatus
            userUrl.text = miniUser.url
            userUrl.isClickable = true
            Linkify.addLinks(userUrl, Linkify.WEB_URLS)


            val profileImageReference = mStorageReference.child("images/users/" + miniUser.userId)
            profileImageReference.downloadUrl.addOnSuccessListener { uri ->
                Picasso.with(mContext)
                        .load(uri)
                        .resize(148, 148)
                        .centerCrop()
                        .transform(CircleTransform())
                        .into(userThumb)
            }.addOnFailureListener {
                // Handle any errors
                userThumb.setImageDrawable(mContext.getDrawable(R.drawable.ic_profile_black_24dp))
            }

            rowView.setOnClickListener {
                val intent = Intent(mContext, UserDetailActivity::class.java)
                intent.putExtra("userId", miniUser.userId)
                mContext.startActivity(intent)
            }

            return rowView
        }
    }
}
